index.js

```diff
@@ -0,0 +1,60 @@
+'use strict';
+
+const renderers = require('./lib/renderers');
+const autoStopReporting = require('./lib/autoStopReporting');
+
+function reporter(compiler, options) {
+    options = Object.assign({
+        stats: true,
+
+        write: (str) => str && process.stderr.write(str),
+
+        printStart: (renderStart) => renderStart(),
+        printSuccess: (renderSuccess, ...rest) => renderSuccess(...rest),
+        printError: (renderError, ...rest) => renderError(...rest),
+        printStats: (renderStats, ...rest) => renderStats(...rest),
+    }, options);
+
+    let displayStats = options.stats === true || options.stats === 'once';
+
+    function onStart() {
+        options.write(options.printStart(renderers.start));
+    }
+
+    function onSuccess(stats) {
+        options.write(options.printSuccess(renderers.success, stats.endTime - stats.startTime));
+
+        if (displayStats) {
+            displayStats = options.stats === 'once' ? false : displayStats;
+
+            options.write(options.printStats(renderers.stats, stats));
+        }
+    }
+
+    function onError(err) {
+        options.write(options.printError(renderers.error, err));
+
+        if ((err.stats && err.stats.hasErrors && err.stats.hasErrors()) && displayStats) {
+            displayStats = options.stats === 'once' ? false : displayStats;
+
+            options.write(options.printStats(renderers.stats, err.stats));
+        }
+    }
+
+    function stopReporting() {
+        compiler
+        .removeListener('begin', onStart)
+        .removeListener('end', onSuccess)
+        .removeListener('error', onError);
+    }
+
+    options.printStart && compiler.on('begin', onStart);
+    options.printSuccess && compiler.on('end', onSuccess);
+    options.printError && compiler.on('error', onError);
+
+    autoStopReporting(compiler, stopReporting);
```

Comment:
Do you think the "automatic" stop is necessary? I was expecting that report would only stop if `stopReporting` was called. Though, the stats option when `once` must work correctly on subsequent calls of `run` or `watch`.